### PR TITLE
 refector: refect the pouch's volume module

### PR DESCRIPTION
**Pull request by rudyfly**
### Ⅰ. Describe what this PR did
1. don't return error when create volume is exist.
2. Refect the `VolumeID` to `VolumeContext`.
3. remove unuse field `Selectors` in `VolumeContext`.
4. set directory's quota id after check the limit of device.
5. get volume's size in `Extra` field.
6. add more options to set volume's size, you also can use "-o size=10g" or
"-o Size=10g" or "-o opt.Size=10g" to set volume's size.
7. merge volume's error into `pkg/errtypes`
8. remove unuse package `pkg/serializer`

### Ⅱ. Does this pull request fix one issue?

NONE

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
Added `TestToStringMap`

### Ⅴ. Special notes for reviews

Signed-off-by: Rudy Zhang <rudyflyzhang@gmail.com>

### CODE REVIEW

File: daemon/mgr/container_storage.go [hunk 1]
```diff
@@ -173,22 +173,29 @@ func (mgr *ContainerManager) getMountPointFromBinds(ctx context.Context, c *Cont
...
 			}
 
-			if mp.Replace != "" {
-				mp.Source, err = mgr.VolumeMgr.Path(ctx, name)
-				if err != nil {
-					return err
-				}
+			volume, err := mgr.VolumeMgr.Get(ctx, name)
+			if err != nil || volume == nil {
+				logrus.Errorf("failed to get volume: (%s), err: (%v)", name, err)
```
Comment (allencloud) [inline daemon/mgr/container_storage.go]: Could you use the same way to construct log and error message? 
I found that in L178, you use `logrus.Errorf("failed to bind volume(%s): %v", name, err)`, but here `logrus.Errorf("failed to get volume: (%s), err: (%v)", name, err)`.
How about that?

File: storage/volume/types/volume.go [hunk 2]
```diff
@@ -193,21 +200,25 @@ func (v *Volume) CreateTime() string {
 
 // VolumeID use to define the volume's identity.
 type VolumeID struct {
```
Comment (allencloud) [inline storage/volume/types/volume.go]: Could we change the name of VolumeID？It is so strange.
Is this that this VolumeID will be stored in some legacy db?

File: storage/quota/quota.go [hunk 3]
```diff
@@ -156,6 +156,21 @@ func GetNextQuotaID() (uint32, error) {
...
+// GetQuotaID returns the quota id of directory,
+// if no quota id, it will alloc the next available quota id.
+func GetQuotaID(dir string) (uint32, error) {
+	id := GetQuotaIDInFileAttr(dir)
+	if id > 0 {
+		return id, nil
+	}
+	id, err := GetNextQuotaID()
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to get file: (%s) quota id", dir)
```
Comment (allencloud) [inline storage/quota/quota.go]: please remove `:`

File: storage/quota/quota.go [hunk 4]
```diff
@@ -181,18 +196,20 @@ func GetDefaultQuota(quotas map[string]string) string {
 func SetRootfsDiskQuota(basefs, size string, quotaID uint32) (uint32, error) {
 	overlayMountInfo, err := getOverlayMountInfo(basefs)
 	if err != nil {
-		return 0, fmt.Errorf("failed to get overlay mount info: %v", err)
+		return 0, errors.Wrapf(err, "failed to get overlay:(%s) mount info", basefs)
```
Comment (allencloud) [inline storage/quota/quota.go]: please remove `:`

File: daemon/mgr/volume.go [hunk 5]
```diff
@@ -83,6 +83,9 @@ func (vm *VolumeManager) Create(ctx context.Context, name, driver string, option
 
 	v, err := vm.core.CreateVolume(id)
 	if err != nil {
+		if e, ok := err.(*volerr.CoreError); ok && e.IsVolumeExisted() {
```
Comment (allencloud) [inline daemon/mgr/volume.go]: Could we have a way to combine `pkg/errtypes` and `storage/volume/error` into one?
Or extract the common part? @rudyfly

File: storage/volume/core.go [hunk 6]
```diff
@@ -159,43 +159,43 @@ func (c *Core) getVolumeDriver(id types.VolumeID) (*types.Volume, driver.Driver,
 }
 
 // existVolume return 'true' if volume be found and not errors.
-func (c *Core) existVolume(id types.VolumeID) (bool, error) {
-	_, err := c.getVolume(id)
+func (c *Core) existVolume(id types.VolumeContext) (*types.Volume, bool, error) {
```
Comment (allencloud) [inline storage/volume/core.go]: I am wondering if it is necessary to update the return parameter of this function. I saw that you added an instance of `*types.Volume` in the function.

When I searched the code, I found that this is used in `core.CreateVolume`, and this `*types.Volume` is only used when the input volume ID exists, then you returned the volume.

I think with the change, you break the primitives of this function.

As an encapsulated package, we should make the function to be focused as much as possible. when checking if it exists, the main part is the returned boolean, if the returned error is necessary, it is ok. 

But when we return a Volume instance, it combines two actions into a single one, to break the purity of this function `existVolume`.

If the caller need the `types.Volume` instance when it is already exists, please let it use another function to get it, like `core.GetVolumes(...)`.

WDYT? @rudyfly 

In addition, would you mind changing function name into `isVolumeExist`?

**Comment by allencloud** [comment]
For user's input size, Size, opt.Size or others, could we first convert them into bytes which is a type of integer. So all the value dealing is all about integer rather rather string.

When to the lower part where we must to convert them into string, then we do the convertion.

Otherwise the string dealing all along the path may lead to some unreadability or some potential bugs. WDYT? @rudyfly

File: daemon/mgr/volume.go [hunk 7]
```diff
@@ -65,12 +66,11 @@ func (vm *VolumeManager) Create(ctx context.Context, name, driver string, option
 		driver = types.DefaultBackend
 	}
 
-	id := types.VolumeID{
-		Name:      name,
-		Driver:    driver,
-		Options:   map[string]string{},
-		Selectors: map[string]string{},
-		Labels:    map[string]string{},
+	id := types.VolumeContext{
```
Comment (HusterWan) [inline daemon/mgr/volume.go]: change `VolumeID` to `VolumeContext`, feel better.

File: storage/volume/types/volume.go [hunk 8]
```diff
@@ -162,9 +155,19 @@ func (v *Volume) SetLabel(label, value string) {
 	v.Labels[label] = value
 }
 
-// Size returns volume's size(MB).
+// Size returns volume's size(bytes).
 func (v *Volume) Size() string {
-	return v.Spec.Size
+	if v.Spec.Size != "" {
```
Comment (HusterWan) [inline storage/volume/types/volume.go]: Can we add some verification for the `Size` in API interface?

**Comment by rudyfly** [comment]
@zhuangqh 
fixed, please see this: https://github.com/alibaba/pouch/pull/2379/files#diff-10877fe186b33251a42f1b8719753b49R205

File: storage/volume/types/volume.go (hunk 8 above)
Comment (rudyfly) [reply storage/volume/types/volume.go]: No need 😄

File: daemon/mgr/volume.go (hunk 7 above)
Comment (rudyfly) [reply daemon/mgr/volume.go]: 🐶

**Comment by rudyfly** [comment]
@allencloud It's hard to change the type of `size` now, because we have store it into boltdb.

File: storage/volume/core.go (hunk 6 above)
Comment (rudyfly) [reply storage/volume/core.go]: good idea, I fix it.

File: daemon/mgr/container_storage.go (hunk 1 above)
Comment (rudyfly) [reply daemon/mgr/container_storage.go]: fixed

File: storage/quota/quota.go (hunk 3 above)
Comment (rudyfly) [reply storage/quota/quota.go]: fixed

**Comment by HusterWan** [comment]
Since I still have some comments, but I think them not stuck for merge this PR, 
So LGTM, thanks for your works, Boss Ku !

**Comment by HusterWan** [comment]
lgtm

**Pull request merged**